### PR TITLE
added support for reading full arrays

### PR DESCRIPTION
**Pull request by jjfalling**

### CODE REVIEW

File: lib/logstash/filters/translate.rb [hunk 1]
```diff
@@ -162,35 +162,69 @@ def filter(event)
...
           end
-        elsif @dictionary.include?(source)
-          event[@destination] = @dictionary[source]
-          matched = true
         end
-      else 
-        translation = source.gsub(Regexp.union(@dictionary.keys), @dictionary)
-        if source != translation
-          event[@destination] = translation.force_encoding(Encoding::UTF_8)
+        if not matched and @fallback
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: Is better not to use ```not``` and other word operators, they tend to be slow, is better to use the ones we know from other programing languages.

File: lib/logstash/filters/translate.rb [hunk 2]
```diff
@@ -162,35 +162,69 @@ def filter(event)
...
-      source = event[@field].is_a?(Array) ? event[@field].first.to_s : event[@field].to_s
-      matched = false
-      if @exact
-        if @regex
-          key = @dictionary.keys.detect{|k| source.match(Regexp.new(k))}
-          if key
-            event[@destination] = @dictionary[key]
+      #check if field is an array
+      if event[@field].is_a?(Array)
+        #field is an array, walk it
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: this code looks to me like a perfect candidate for a method. Is a good ruby practice to have small methods that tend to do only one thing, then testing is easier :smiley_cat:

File: lib/logstash/filters/translate.rb [hunk 3]
```diff
@@ -162,35 +162,69 @@ def filter(event)
...
+          end
+          if not matched and @fallback
+            event[@destination][index] = event.sprintf(@fallback)
+            matched = true
+          end
+          filter_matched(event) if matched or @field == @destination[index]
+	end
+      else
+        #Field is not an array
+        source = event[@field].to_s
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: same thing for this part, wrapping this in a function is a perfect thing ! :+1:

File: lib/logstash/filters/translate.rb [hunk 4]
```diff
@@ -162,35 +162,69 @@ def filter(event)
...
+                matched = true
+              end
+            elsif @dictionary.include?(source)
+              event[@destination][index] = @dictionary[source]
+              matched = true
+            end
+          else
+            translation = source.gsub(Regexp.union(@dictionary.keys), @dictionary)
+            if source != translation
+              event[@destination][index] = translation.force_encoding(Encoding::UTF_8)
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: I wonder if this ```force_encoding```  is really necessary here. can the translation be in non UTF-8? in ruby > 2.0 all internal strings are UTF-8 by default while in  ruby < 2.0 if you add a header like ``` #encoding: utf-8  ```  everything is utf-8 then.

Using this method tend to be a red flag for ruby systems.

**Comment by purbon** [comment]
Can you add test for this behaviour? in every PR having test is encouraged and I would say mandatory :P

**Comment by dadoonet** [comment]
+1 for this feature (pull) request! :)

**Comment by purbon** [comment]
yay! helping out @jjfalling here to be more comfortable before going "live" in the main repo :smile_cat:  I also think this PR will make this super interesting! go for it man!

**Comment by jjfalling** [comment]
Just finished the first whack at getting all of that done, but I found a existing bug with dealing with translated values that I'll look at fixing.

File: lib/logstash/filters/translate.rb [hunk 5]
```diff
@@ -147,6 +148,59 @@ def load_yaml(registering=false)
   end
 
   public
+  def match_event(source, event)
+    return unless source
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: why do you have no source? asking this is important, why return nothing (so nil?) is meaningful?

File: lib/logstash/filters/translate.rb [hunk 6]
```diff
@@ -147,6 +148,59 @@ def load_yaml(registering=false)
...
+  end
+
+  public
+  def determine_field_type(event)
+    #Check if field is an array
+    if event[@field].is_a?(Array)
+      #Field is an array, walk it
+      event[@field].each_with_index do |source, index|
+        index = index.to_i
+        matched = false
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: why do you initialize this value here? then you are assigning it later on.

File: lib/logstash/filters/translate.rb [hunk 7]
```diff
@@ -147,6 +148,59 @@ def load_yaml(registering=false)
...
+        event[@destination] ||= []
+        event_item,matched = match_event(source.to_s, event)
+        event[@destination][index] = event_item if matched
+      end
+
+    else
+      #Field is not an array
+      matched = false
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: this should be initialized into the match event function

File: lib/logstash/filters/translate.rb [hunk 8]
```diff
@@ -147,6 +148,59 @@ def load_yaml(registering=false)
...
+    return event_item, matched
+  end
+
+  public
+  def determine_field_type(event)
+    #Check if field is an array
+    if event[@field].is_a?(Array)
+      #Field is an array, walk it
+      event[@field].each_with_index do |source, index|
+        index = index.to_i
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: this is not necessary, see http://ruby-doc.org/core-2.2.0/Enumerable.html#method-i-each_with_index

File: lib/logstash/filters/translate.rb [hunk 9]
```diff
@@ -157,40 +211,19 @@ def filter(event)
...
-        end
-      end
+      determine_field_type(event)
 
-      if not matched and @fallback
-        event[@destination] = event.sprintf(@fallback)
-        matched = true
-      end
-      filter_matched(event) if matched or @field == @destination
     rescue Exception => e
```
Comment (purbon) [inline lib/logstash/filters/translate.rb]: is better to catch standarderror, or use something like:

```
begin
...
rescue => e
end
```

**Pull request closed without merging**